File: src/models/part.ts
import CoursePartItem from "./items/item";
import Course from "./course";
import TextPartItem from "./items/text";
import VideoPartItem from "./items/video";
import QuizPartItem from "./items/quiz";
import CoursesServer from "./server";

export default class CoursePart {
  public readonly name: string;

  public readonly description: string;

  public readonly slug: string;

  public readonly course: Course;

  public readonly items: CoursePartItem[];

  public get server(): CoursesServer {
    return this.course.server;
  }

  public constructor(init?: Partial<CoursePart>) {
    Object.assign(this, init);
    this.items = (this.items as any[]).map((item, index) => {
      item.part = this;
      item.index = index;
      switch ((item.type as string)?.toLowerCase()) {
        case "text":
          return new TextPartItem(item);
        case "video":
          return new VideoPartItem(item);
        case "quiz":
          return new QuizPartItem(item);
        default:
          return null;
      }
    });
  }
}

File: lib/models/part.dart
import 'server.dart';
import 'course.dart';
import 'item.dart';

class CoursePart {
  final Course? course;
  final String? name;
  final String? description;
  final String slug;
  final List<PartItem> items;

  CoursePart(
      {this.name,
      this.description,
      required this.slug,
      List<PartItem> items = const [],
      this.course})
      : this.items = List<PartItem>.unmodifiable(items);
  CoursePart.fromJson(Map<String, dynamic> json)
      : course = json['course'],
        description = json['description'],
        name = json['name'] ?? '',
        slug = json['slug'],
        items = (json['items'] as List<dynamic>? ?? [])
            .map((item) => PartItemTypesExtension.fromName(item['type'])
                .fromJson(Map<String, dynamic>.from(item)))
            .toList();
  Map<String, dynamic> toJson() => {
        "description": description,
        "slug": slug,
        "name": name ?? '',
        "items": items.map((e) => e.toJson()).toList()
      };

  CoursesServer? get server => course!.server;

  CoursePart copyWith(
          {Course? course,
          String? name,
          String? description,
          String? slug,
          List<PartItem>? items}) =>
      CoursePart(
          course: course ?? this.course,
          description: description ?? this.description,
          items: items ?? this.items,
          name: name ?? this.name,
          slug: slug ?? this.slug);
}
